interfaces/prompting/requestrules/requestrules.go

```diff
@@ -41,64 +41,39 @@ import (
+	Interface   string                     `json:"interface"`
+	Constraints *prompting.RuleConstraints `json:"constraints"`
 }
 
-// Validate verifies internal correctness of the rule
-func (rule *Rule) validate(currTime time.Time) error {
-	if err := rule.Constraints.ValidateForInterface(rule.Interface); err != nil {
-		return err
-	}
-	if _, err := rule.Outcome.AsBool(); err != nil {
-		return err
-	}
-	if rule.Lifespan == prompting.LifespanSingle {
-		// We don't allow rules with lifespan "single"
-		return prompting_errors.NewRuleLifespanSingleError(prompting.SupportedRuleLifespans)
-	}
-	if err := rule.Lifespan.ValidateExpiration(rule.Expiration, currTime); err != nil {
-		// Should never error due to an API request, since rules are always
-		// added via the API using duration, rather than expiration.
-		// Error may occur when validating a rule loaded from disk.
-		return err
-	}
-	return nil
+// Validate verifies internal correctness of the rule's constraints and
+// permissions and prunes any expired permissions. If all permissions are
+// expired, then returns true. If the rule is invalid, returns an error.
+func (rule *Rule) validate(currTime time.Time) (expired bool, err error) {
+	return rule.Constraints.ValidateForInterface(rule.Interface, currTime)
 }
 
-// Expired returns true if the receiving rule has a lifespan of timespan and
-// the current time is after the rule's expiration time.
-func (rule *Rule) Expired(currentTime time.Time) bool {
-	switch rule.Lifespan {
-	case prompting.LifespanTimespan:
-		if currentTime.After(rule.Expiration) {
-			return true
-		}
-		// TODO: add lifespan session
-		//case prompting.LifespanSession:
-		// TODO: return true if the user session has changed
-	}
-	return false
+// expired returns true if all permissions for the receiving rule have expired.
+func (rule *Rule) expired(currTime time.Time) bool {
+	return rule.Constraints.Permissions.Expired(currTime)
 }
 
 // variantEntry stores the actual pattern variant struct which can be used to
-// match paths, and the set of rule IDs whose path patterns render to this
-// variant. All rules in a particular entry must have the same outcome, and
-// that outcome is stored directly in the variant entry itself.
+// match paths, and a map from rule IDs whose path patterns render to this
+// variant to the relevant permission entry from that rule. All non-expired
+// permission entry values in the map must have the same outcome (as long as
+// the entry has not expired), and that outcome is also stored directly in the
+// variant entry itself.
 //
 // Use the rendered string as the key for this entry, since pattern variants
 // cannot otherwise be easily checked for equality.
 type variantEntry struct {
-	Variant patterns.PatternVariant
-	Outcome prompting.OutcomeType
-	RuleIDs map[prompting.IDType]bool
+	Variant     patterns.PatternVariant
+	Outcome     prompting.OutcomeType
+	RuleEntries map[prompting.IDType]*prompting.RulePermissionEntry
```

Comment:
why is RulePermissionEntry defined outside of here?

Reply:
A `RulePermissionMap` is a map from permission (string) to `RulePermissionEntry`, so to avoid a circular dependency, `RulePermissionEntry` must be defined in the same place `RulePermissionMap` is (or in a lower level). But at the same time, `RuleConstraints` contains a `RulePermissionMap`, so `RulePermissionMap` must be defined in the same place or lower in the chain than `RuleConstraints`.

An argument could be made for defining `RulePermissionMap` and `RuleConstraints` in `requestrules` instead of `interfaces/prompting`. But `requestprompts` uses `RuleConstraints` as well, so then `requestprompts` would need to import `requestrules`, which is fine but a bit weird to me.

More importantly though, `Constraints` are validated by being converted into `RuleConstraints`. So wherever `Constraints` is defined needs to import whatever defined `RuleConstraints`. So I think `RuleConstraints`, `RulePermissionMap`, and `RulePermissionEntry` all need to be defined in `interfaces/prompting`.

An alternative could be to have separate validation and conversion functions, but these would be essentially identical (and the latter would need to call the former again, probably), so it seems to me like the code overlap would be great, and the conversion functions defined in `requestrules` would need to operate on the internals of those structs as defined in `interfaces/prompting`. So I think it's best to have all the similarly structured and related types defined in one place.